src/wallet/wallet.cpp

```diff
@@ -1359,6 +1359,308 @@ void CWallet::SyncMetaData(pair<typename TxSpendMap<T>::iterator, typename TxSpe
+        [&](const libzcash::SaplingPaymentAddress& addr) {
+            auto meta = GetUFVKMetadataForReceiver(addr);
+            if (meta.has_value()) {
+                result = self->GetUnifiedAccountId(meta.value().first);
+            }
+        },
+        [&](const AccountZTXOPattern& acct) {
+            if (self->mnemonicHDChain.has_value() &&
+                self->mapUnifiedAccountKeys.count(std::make_pair(self->mnemonicHDChain.value().GetSeedFingerprint(), acct.GetAccountId())) > 0) {
+                result = acct.GetAccountId();
+            }
+        }
+    }, selector.GetPattern());
+    return result;
+}
+
+bool CWallet::SelectorMatchesAddress(
+        const ZTXOSelector& selector,
+        const CTxDestination& address) const {
+    auto self = this;
+    return std::visit(match {
+        [&](const CKeyID& keyId) {
+            CTxDestination keyIdDest = keyId;
+            return address == keyIdDest;
+        },
+        [&](const CScriptID& scriptId) {
+            CTxDestination scriptIdDest = scriptId;
+            return address == scriptIdDest;
+        },
+        [&](const libzcash::SproutPaymentAddress& addr) { return false; },
+        [&](const libzcash::SaplingPaymentAddress& addr) { return false; },
+        [&](const AccountZTXOPattern& acct) {
+            if (acct.IncludesP2PKH() || acct.IncludesP2SH()) {
+                std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>> meta;
+                std::visit(match {
+                    [&](const CNoDestination& none) { meta = std::nullopt; },
+                    [&](const auto& addr) { meta = self->GetUFVKMetadataForReceiver(addr); }
+                }, address);
+                if (meta.has_value()) {
+                    // use the coin if the account id corresponding to the UFVK is
+                    // the payment source account.
+                    return self->GetUnifiedAccountId(meta.value().first) == std::optional(acct.GetAccountId());
+                } else {
+                    // The legacy account is treated as a single pool of
+                    // transparent funds, reproducing wallet behavior prior to
+                    // the advent of unified addresses.
+                    return acct.GetAccountId() == ZCASH_LEGACY_ACCOUNT;
+                }
+            }
+            return false;
+        }
+    }, selector.GetPattern());
+}
+
+bool CWallet::SelectorMatchesAddress(
+        const ZTXOSelector& selector,
+        const libzcash::SproutPaymentAddress& a0) const {
+    return std::visit(match {
+        [&](const CKeyID& keyId) { return false; },
+        [&](const CScriptID& scriptId) { return false; },
+        [&](const libzcash::SproutPaymentAddress& a1) { return a0 == a1; },
+        [&](const libzcash::SaplingPaymentAddress& addr) { return false; },
+        [&](const AccountZTXOPattern& acct) { return false; }
+    }, selector.GetPattern());
```

Comment:
```suggestion
    auto a1 = std::get<libzcash::SproutPaymentAddress&>(selector.GetPattern());
    return a1 != std::nullopt && a0 == a1.value();
```

Reply:
`std::get` returns a nullable pointer, not a `std::optional` value.